Update Using AWS Inferentia to run HuggingFace TorchScript model (#14771)
huggingface/transformers: docs/source/serialization.rst

custom-built hardware accelerator, specializing in deep learning inferencing workloads. `AWS Neuron
<https://awsdocs-neuron.readthedocs-hosted.com/en/latest/#>`_ is the SDK for Inferentia that supports tracing and
optimizing transformers models for deployment on Inf1. The Neuron SDK provides:

1. Easy-to-use API with one line of code change to trace and optimize a TorchScript model for inference in the cloud.
2. Out of the box performance optimizations for `improved cost-performance
   <https://awsdocs-neuron.readthedocs-hosted.com/en/latest/neuron-guide/benchmark/>`_.
3. Support for HuggingFace transformers models built with either `PyTorch
   <https://awsdocs-neuron.readthedocs-hosted.com/en/latest/src/examples/pytorch/bert_tutorial/tutorial_pretrained_bert.html>`_
   or `TensorFlow
   <https://awsdocs-neuron.readthedocs-hosted.com/en/latest/src/examples/tensorflow/huggingface_bert/huggingface_bert.html>`_.

Implications
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Transformers Models based on the `BERT (Bidirectional Encoder Representations from Transformers)

> <https://huggingface.co/bert-base-uncased>`_ architecture, or its variants such as `distilBERT

Reviewer: This link is still wrong

Author: HI @philschmid thank you for the suggestion, the link should be consistent with that of other models now. I updated BERT link to https://huggingface.co/docs/transformers/master/model_doc/bert

Using the traced model for inference is as simple as using its ``__call__`` dunder method:

.. code-block:: python                                              

    traced_model(tokens_tensor, segments_tensors)                   

> Deploying HuggingFace TorchScript models on AWS using the Neuron SDK

Reviewer: ```suggestion
Converting HuggingFace Transformers to AWS Neuron compatible models
```

custom-built hardware accelerator, specializing in deep learning inferencing workloads. `AWS Neuron
<https://awsdocs-neuron.readthedocs-hosted.com/en/latest/#>`_ is the SDK for Inferentia that supports tracing and
optimizing transformers models for deployment on Inf1. The Neuron SDK provides:

1. Easy-to-use API with one line of code change to trace and optimize a TorchScript model for inference in the cloud.
2. Out of the box performance optimizations for `improved cost-performance
   <https://awsdocs-neuron.readthedocs-hosted.com/en/latest/neuron-guide/benchmark/>`_.
3. Support for HuggingFace transformers models built with either `PyTorch
   <https://awsdocs-neuron.readthedocs-hosted.com/en/latest/src/examples/pytorch/bert_tutorial/tutorial_pretrained_bert.html>`_
   or `TensorFlow
   <https://awsdocs-neuron.readthedocs-hosted.com/en/latest/src/examples/tensorflow/huggingface_bert/huggingface_bert.html>`_.

Implications
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Transformers Models based on the `BERT (Bidirectional Encoder Representations from Transformers)

> <https://huggingface.co/bert-base-uncased>`_ architecture, or its variants such as `distilBERT

Reviewer: Can you please align those documentation links to the newest documentation, e.g. `distilbert` -> https://huggingface.co/docs/transformers/master/model_doc/distilbert

<https://huggingface.co/transformers/v4.0.1/model_doc/distilbert.html>`_ and `roBERTa

> <https://huggingface.co/roberta-base>`_ will run best on Inf1 for non-generative tasks such as Extractive Question

Reviewer: Can you please align those documentation links to the newest documentation, e.g. `distilbert` -> https://huggingface.co/docs/transformers/master/model_doc/distilbert

Author: Thank you and it is changed according to your suggestion.

Answering, Sequence Classification, Token Classification. Alternatively, text generation tasks can be adapted to run on
Inf1, according to this `AWS Neuron MarianMT tutorial
<https://awsdocs-neuron.readthedocs-hosted.com/en/latest/src/examples/pytorch/transformers-marianmt.html>`_. More
information about models that can be converted out of the box on Inferentia can be found in the `Model Architecture Fit
section of the Neuron documentation
<https://awsdocs-neuron.readthedocs-hosted.com/en/latest/neuron-guide/models/models-inferentia.html#models-inferentia>`_.

Dependencies
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

> Using AWS Neuron to run inference workloads requires the following dependencies and environment:

Reviewer: ```suggestion
Using AWS Neuron to convert models requires the following dependencies and environment:
```

> 1. An `Amanzon EC2 Inf1 <https://aws.amazon.com/ec2/instance-types/inf1/>`_ instance in AWS cloud.

Reviewer: ```suggestion
```

> 2. A `Neuron SDK environment

Reviewer: ```suggestion
* A `Neuron SDK environment
```

   <https://awsdocs-neuron.readthedocs-hosted.com/en/latest/neuron-guide/neuron-frameworks/pytorch-neuron/index.html#installation-guide>`_,
   which comes pre-configured on `AWS Deep Learning AMI
   <https://docs.aws.amazon.com/dlami/latest/devguide/tutorial-inferentia-launching.html>`_

> Running Neuron traced TorchScript in Inf1

Reviewer: ```suggestion
Converting a Model for AWS Neuron
```

.. code-block:: python

    traced_model(tokens_tensor, segments_tensors)

Using HuggingFace TorchScript model in AWS Inf1 using Neuron SDK
-----------------------------------------------------------------------------------------------------------------------

> AWS introduced `Inf1 <https://aws.amazon.com/ec2/instance-types/inf1/>`_ for low cost, high performance machine learning model inference tasks in AWS cloud. 

Reviewer: We should think about calling inferentia: `inf1` or `inferentia` in the whole document and not mix them up. This can confuse none aws experienced people

Author: Agreed and all should be Inf1. Changes made in new commit.

Author: Appreciate the suggestion and changes are made where appropriate. We used the term 'Inf1' to refer to EC2 instance type, and 'Inferentia' to refer to the chip name. Rebased and resubmit in a new PR: https://github.com/huggingface/transformers/pull/14982

learning workloads in Inf1.` In Neuron SDK, there are these benefhts:

1. Easy-to-use API with one line of code change to trace and optimize a TorchScript model for inference in the cloud.
2. Out of the box performance optimizations for `improved cost-performance <https://awsdocs-neuron.readthedocs-hosted.com/en/latest/neuron-guide/benchmark/>`_ using the AWS Neuron.
3. Support for HuggingFace (HF) models built with either `PyTorch <https://awsdocs-neuron.readthedocs-hosted.com/en/latest/src/examples/pytorch/bert_tutorial/tutorial_pretrained_bert.html>`_ or `TensorFlow <https://awsdocs-neuron.readthedocs-hosted.com/en/latest/src/examples/tensorflow/huggingface_bert/huggingface_bert.html>`_.

Implications
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Models with transformer based architecture, i.e., 
`BERT (Bidirectional Encoder Representations from Transformers) <https://huggingface.co/bert-base-uncased>`_ or its variants
such as `distilBERT <https://huggingface.co/transformers/v4.0.1/model_doc/distilbert.html>`_ 

> and `roBERTa <https://huggingface.co/roberta-base>`_ will run best in Inf1 for non-generative tasks such as Extractive Question Answering, Sequence Classification, NER. 

Reviewer: We should align the ref links to those models. I think adding the `docs` link for all 3 is probably better than the models weights on the hub or what do you think?

Author: Agreed and resolved. .DS_Store file removed, and links are all referring to the master docs links. Rebased and resubmitted to a new PR: https://github.com/huggingface/transformers/pull/14982

> and `roBERTa <https://huggingface.co/roberta-base>`_ will run best in Inf1 for non-generative tasks such as Extractive Question Answering, Sequence Classification, NER. 

Reviewer: ```suggestion
and `roBERTa <https://huggingface.co/roberta-base>`_ will run best in Inf1 for non-generative tasks such as Extractive Question Answering, Sequence Classification, Token Classification. 
```

Alternatively, text generation tasks can be adapted to run on Inf1, according to this tutorial using `MarianMT <https://awsdocs-neuron.readthedocs-hosted.com/en/latest/src/examples/pytorch/transformers-marianmt.html>`_.

> More information about models that perform well out of the box on Inferentia can be found in the `Model Architecture Fit section of the Neuron documentation <https://awsdocs-neuron.readthedocs-hosted.com/en/latest/neuron-guide/models/models-inferentia.html#models-inferentia>`_.

Reviewer: ```suggestion
More information about models that can be converted out of the box for Inf1 can be found in the `Model Architecture Fit section of the Neuron documentation <https://awsdocs-neuron.readthedocs-hosted.com/en/latest/neuron-guide/models/models-inferentia.html#models-inferentia>`_.
```
